---
title: "class10:Structural Bioinformatics"
author: "Yufei Zhang A16987415"
date: "`r Sys.Date()`"
format: pdf
---

1: Introduction to the RCSB Protein Data Bank (PDB)
The PDB archive is the major repository of information about the 3D structures of large biological molecules, including proteins and nucleic acids. 

```{r}
pdb_data<-read.csv('Data Export Summary.csv')
```

```{r}
library(dplyr)

pdb_data <- pdb_data %>%
  mutate(across(.cols = -1, .fns = ~ case_when(
    is.character(.x) ~ as.numeric(gsub("[^0-9.-]", "", .x)),  # Clean and convert to numeric
    TRUE ~ as.numeric(.x)  # Default to convert to numeric if possible
  )))
```

```{r}
pdb_data
```
Q1: What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.

```{r}
total<-sum(pdb_data$Total)
xsum<-sum(pdb_data$X.ray)
xpercentage<-xsum/total
xpercentage
```
84% are x-ray.

```{r}
ESum<-sum(pdb_data$EM)
Epercentage<-ESum/total
Epercentage
```
9.2% percent are EM.

Q2: What proportion of structures in the PDB are protein?


```{r}
protein<-pdb_data[1,'Total']
proteinperc<-protein/total
proteinperc
```
86.59% are protein.

Q3: Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

4445 Structures.

Q4: Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?

It only displays the oxygen atom and each water molecule represents one residue.

Q5: There is a critical “conserved” water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have

It have #308 residue number.

Q6: Generate and save a figure clearly showing the two distinct chains of HIV-protease along with the ligand. You might also consider showing the catalytic residues ASP 25 in each chain and the critical water (we recommend “Ball & Stick” for these side-chains). Add this figure to your Quarto document.

![](1HSG.png)

Discussion Topic: Can you think of a way in which indinavir, or even larger ligands and substrates, could enter the binding site?

Q7: [Optional] As you have hopefully observed HIV protease is a homodimer (i.e. it is composed of two identical chains). With the aid of the graphic display can you identify secondary structure elements that are likely to only form in the dimer rather than the monomer?

Alpha helices.

3. Introduction to Bio3D in R
Bio3D is an R package for structural bioinformatics. Features include the ability to read, write and analyze biomolecular structure, sequence and dynamic trajectory data.
```{r}
library(bio3d)

```
```{r}
pdb <- read.pdb("1hsg")
```

```{r}
pdb
```
Q7: How many amino acid residues are there in this pdb object? 

198 residues.

Q8: Name one of the two non-protein residues? 

HOH, water molecule.

Q9: How many protein chains are in this structure? 

There are 2 protein chains.

```{r}
attributes(pdb)
```
```{r}
head(pdb$atom)
```

Predicting functional motions of a single structure
```{r}
adk <- read.pdb("6s36")
```
```{r}
adk
```
```{r}
# Perform flexiblity prediction
m <- nma(adk)
```
```{r}
plot(m)
```
```{r}
mktrj(m, file="adk_m7.pdb")
```

4. Comparative structure analysis of Adenylate Kinase
The goal of this section is to perform principal component analysis (PCA) on the complete collection of Adenylate kinase structures in the protein data-bank (PDB).

```{r}
# Install packages in the R console NOT your Rmd/Quarto file

#install.packages("bio3d")
#install.packages("devtools")
#install.packages("BiocManager")

#BiocManager::install("msa")
#devtools::install_bitbucket("Grantlab/bio3d-view")
```

Q10. Which of the packages above is found only on BioConductor and not CRAN? 

The msa package.

Q11. Which of the above packages is not found on BioConductor or CRAN?: 

bio3d-view.

Q12. True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket? 
 
 True
 
Search and retrieve ADK structures
```{r}
library(bio3d)
aa <- get.seq("1ake_A")
```
```{r}
aa
```
Q13. How many amino acids are in this sequence, i.e. how long is this sequence? 

214 amino acids.

```{r}
hits <- NULL
hits$pdb.id <- c('1AKE_A','6S36_A','6RZE_A','3HPR_A','1E4V_A','5EJE_A','1E4Y_A','3X2S_A','6HAP_A','6HAM_A','4K46_A','3GMT_A','4PZL_A')
# Download releated PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

Align and superpose structures

```{r}
library(msa)

# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
```
```{r}
# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment
#plot(pdbs, labels=ids)

```

```{r}
anno <- pdb.annotate(ids)
unique(anno$source)
```

Principal component analysis
```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

#plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```


```{r}
# Visualize first principal component
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```

```{r}
#Plotting results with ggplot2
library(ggplot2)
library(ggrepel)
df <- data.frame(PC1=pc.xray$z[,1], 
                 PC2=pc.xray$z[,2], 
                 col=as.factor(grps.rd),
                 ids=ids)

p <- ggplot(df) + 
  aes(PC1, PC2, col=col, label=ids) +
  geom_point(size=2) +
  geom_text_repel(max.overlaps = 20) +
  theme(legend.position = "none")
p
```
6. Normal mode analysis [optional]
```{r}
# NMA of all structures
modes <- nma(pdbs)
plot(modes, pdbs, col=grps.rd)
```

Q14. What do you note about this plot? Are the black and colored lines similar or different? Where do you think they differ most and why?

Black and colored lines are very different. They differ the most from residue 125-150 since they are ligand binding sites and the confirmation changes a lot with ligand binding or not binding

